[#121933113] User-Agent frontend

saliceti opened the pull request:
# What

Story: [Fix Router issue for zero downtime deploy](https://www.pivotaltracker.com/story/show/121933113)

[Gorouter](https://github.com/cloudfoundry/gorouter) provides a special healthcheck response when a request has the 'User-Agent: HTTP-Monitor/1.1' HTTP header. Unfortunately ELBs don't send this header.
This PR adds a HAProxy frontend to set this header on requests to the default_backend (in our case gorouter).
# How to review

It can be tricky to test in isolation. I suggest to test with the paas-cf PR that includes this one.
# After merge

Create a new tag on the merge commit and update the relevant commit in the paas-cf PR.
# Who can review

Anyone but @combor, @timmow  or myself.

saliceti committed:
Add monitoring frontend with custom User-Agent

In order to allow AWS ELBs to monitor gorouter and detect when it is
draining connections, we add an additional frontend on an alternate port
which inserts a User Agent [1] header on the request to backend.
Gorouter identifies this header, responding 503 if it is draining connections.
This way ELB will detect the node as unhealthy and remove it from the pool of
backends, so it will stop sending new connections.

[1] https://github.com/cloudfoundry/gorouter#instrumentation

keymon commented:
In the description of the new variable `ha_proxy.user_agent_frontend` you don't say which are the valid values for it. Additionally the name is not really meaningful.

Reading the logic, assigning any value will enable this feature, and will open the port 82 sending a hardcoded value in the UserAgent.

I suggest that you either:
 * Change the name to `ha_proxy.enable_monitor_healthcheck_frontend`, with a description description to explain that is "true" or "false", that will open the port, and default it to false.
 * Or call it to something like:
    * `ha_proxy.healthcheck_frontend.enabled` = true|false, default: false
    * `ha_proxy.healthcheck_frontend.user_agent` default: `HTTP-Monitor/1.1`

keymon commented:
Can you explain it better in the commit.

For example:

>  Add monitoring frontend with custom User-Agent 
> 
> In order to allow AWS ELBs monitor the gorouter and detect when is draining connections, we add a additional frontend in an alternate port which will send a custom User Agent [1]. GoRouter identifies this custom agent, responding 503 if it is draining connections. 
> This way ELB will detect the node as bad and remove it from the pool of backends, so it will stop sending new connections.
>
> [1] http://link_to_code_in_go_rotuer

Adapt to explain the new options in the spec if required.
requests to backend.
Especially useful for gorouter healthcheck.

dcarley commented on jobs/haproxy/spec:
```diff
@@ -50,6 +50,10 @@ properties:
     default: 80
   ha_proxy.additional_frontend_config:
     description: "Configurations which are passed to the http-in and https-in configuration block"
+  ha_proxy.user_agent_frontend:
```
From meatspace: we'll rename to `enable_healthcheck_frontend` to describe the functionality rather than the implementation.

saliceti committed:
Fix disable_http property test

p(“variable”) always evaluate to true so it wasn’t possible to enable
http. Now we compare the strings for the test to work.

keymon commented on jobs/haproxy/templates/haproxy.conf.erb:
```diff
@@ -72,6 +72,16 @@ frontend ssl-in
     default_backend tcp-routers
 <% end %>
 
+<% if p("ha_proxy.enable_healthcheck_frontend") == "true" do %>
```
The `do` is NOT needed for the if...

```

Error 100: Unable to render instance groups for deployment. Errors are:
   - Unable to render jobs for instance group 'router'. Errors are:
     - Unable to render templates for job 'haproxy'. Errors are:
       - Error filling in template '(unknown)' (line (unknown): haproxy/haproxy.conf.erb:75: syntax error, unexpected keyword_do, expecting keyword_then or ';' or '\n'
      ...thcheck_frontend") == "true" do ; _erbout.concat "\nfrontend...
      ...                               ^
      haproxy/haproxy.conf.erb:83: syntax error, unexpected keyword_end, expecting end-of-input
      ;  end ; _erbout.concat "\n\nbacken...
            ^)
       - Error filling in template '(unknown)' (line (unknown): haproxy/haproxy.ctmpl.erb:75: syntax error, unexpected keyword_do, expecting keyword_then or ';' or '\n'
      ...thcheck_frontend") == "true" do ; _erbout.concat "\nfrontend...
      ...                               ^
      haproxy/haproxy.ctmpl.erb:83: syntax error, unexpected keyword_end, expecting end-of-input
      ;  end ; _erbout.concat "\n\nbacken...
            ^)

```

:(

keymon commented:
I fixed the `if ...  do ` issue

keymon commented:
Oh! unfortunately the assumption that the `p()` function returns a string "true" or "false" as Strings is wrong.

You can test it with:

```
aws s3 cp s3://hector-state/cf-manifest.yml /tmp/cf-manifest.yml --region eu-west-1 
./platform-tests/bosh-template-renderer/render.rb  ../paas-haproxy-release/jobs/haproxy/templates/haproxy.ctmpl.erb ../paas-haproxy-release/jobs/haproxy/spec /tmp/cf-manifest.yml
```

My bad I did not test this properly after the latest changes, I assumed that was working. We need some tests around it :(
